src/v/raft/consensus.cc

```diff
@@ -1907,31 +1911,49 @@ ss::future<install_snapshot_reply> consensus::finish_snapshot(
 }
 
 ss::future<> consensus::write_snapshot(write_snapshot_cfg cfg) {
-    return _op_lock.with([this, cfg = std::move(cfg)]() mutable {
-        // do nothing, we already have snapshot for this offset
-        // MUST be checked under the _op_lock
-        if (cfg.last_included_index <= _last_snapshot_index) {
-            return ss::now();
-        }
-        auto max_offset = cfg.should_truncate ? _commit_index
-                                              : last_visible_index();
+    return _op_lock
+      .with([this, cfg = std::move(cfg)]() mutable {
+          // do nothing, we already have snapshot for this offset
+          // MUST be checked under the _op_lock
+          if (cfg.last_included_index <= _last_snapshot_index) {
+              return ss::now();
+          }
 
-        vassert(
-          cfg.last_included_index <= max_offset,
-          "Can not take snapshot, requested offset: {} is greater than max "
-          "snapshot offset: {}",
-          cfg.last_included_index,
-          _commit_index);
+          auto max_offset = last_visible_index();
+          vassert(
+            cfg.last_included_index <= max_offset,
+            "Can not take snapshot, requested offset: {} is greater than max "
+            "snapshot offset: {}",
+            cfg.last_included_index,
+            max_offset);
 
-        return do_write_snapshot(cfg.last_included_index, std::move(cfg.data))
-          .then([this, should_truncate = cfg.should_truncate] {
-              if (!should_truncate) {
-                  return ss::now();
-              }
-              return truncate_to_latest_snapshot();
-          })
-          .then([this] { _log.set_collectible_offset(_last_snapshot_index); });
-    });
+          return do_write_snapshot(
+            cfg.last_included_index, std::move(cfg.data));
+      })
+      .then([this] {
+          // Release the lock when truncating the log because it can take some
+          // time while we wait for readers to be evicted.
+          return _log.truncate_prefix(storage::truncate_prefix_config(
+            details::next_offset(_last_snapshot_index),
+            _scheduling.default_iopc));
+      })
+      .then([this] {
+          return _op_lock.with([this] {
```

Comment:
since we release an `_oplock`, it may be possible that this continuation will be reordered if there were two `write_snapshot` methods pending. Is calling `offset_translator` and `configuration_manager` `prefix_truncate` member function safe with stale offset  i.e. the one that was already prefix truncated ?

Reply:
This is not a problem because `_last_snapshot_index` only moves forward and calling `prefix_truncate` with a stale offset is a no-op.

But another case - prefix truncating with a *newer* offset might be a problem because we can prefix truncate translator before prefix truncating the log (which defeats the whole purpose of the patch). So I guess we are better off with saving the snapshot index after writing the snapshot and using it with later continuations.